---
output:
knitrBootstrap::bootstrap_document:
title: Response to Buettner, Natarajan "Computational analysis of cell-to-cell heterogeneity in single-cell RNA-sequencing data reveals hidden subpopulations of cells"
author: 
- Andrew McDavid
- Greg Finak
- Raphael Gottardo
- Fred Hutchinson Cancer Research Center
date:  "`r format(Sys.time(), '%d %B, %Y')`"
keep_md: yes
---
<!-- using knitrBootstrap::knit_bootstrap -->
```{r libraries, echo=TRUE, error=FALSE, cache=FALSE, echo=FALSE,warning=FALSE}
knitr::opts_chunk$set(message=FALSE, error=FALSE, warning=FALSE, cache=TRUE, auto.dep=TRUE, echo=TRUE, tidy=TRUE, width.cutoff=50)
options(digits=2)
suppressPackageStartupMessages({
library(data.table)
library(ggplot2)
library(GGally)
library(gdata)
library(gridExtra)
library(org.Mm.eg.db)
library(limma)
stopifnot(packageVersion('limma')>='3.22.5') #needed because function ids2indices was recently renamed
library(GSEABase)
library(Biobase)
library(stringr)
library(Rtsne)
library(MASS)
library(reshape2)
})
```

## Loading mESC data
We first load the expression data that the authors provide in the paper, although note that the final gene in the mESC data set contained corrupted data in the file included in the Nature supplement.   John Marioni kindly provided us with a version that fixed that corruption, which is what is being used here.

We also load the list of cell cycle genes, and an estimate of the Panama (scLVM) latent variable.  In the remainder of this document we refer to this variable as just the panama latent variable, since that is what BNC are doing in the python code provided.
```{r loadmesc}
## Supplementary table 1, sheet 1
ccgenes <- read.xls('data/cell_cycle_genes.xlsx', as.is=TRUE, header=F)$V1
## converted from excel .xlsx, Supplementary Data 2, fixing a corrupted character in final column.
##These are Log10(ercc_normalized_counts+1).
uncorr <- fread('data/mesc_cell_cycle_uncorrected.csv')
## Generated from above using authors' code from https://github.com/PMBio/scLVM
Xpanama <- scan('data/ES_X_ARD_1.txt')
mesc <- as.matrix(uncorr[,-1,with=FALSE])
## using totals of each phase listed in paper
mescCdat <- data.frame(cycle_class=factor(c(rep("G1",59),  rep("S", 58),  rep("G2M",65)), levels=c('G1', 'S', 'G2M')), idx=seq_len(nrow(mesc)), panama=Xpanama, geomsizeUncorr=rowSums(mesc))
```

We'll also need a crosswalk between the gene symbols and ensembl ids provided in `ccgenes`:
```{r}
eg <- as.data.table(as.data.frame(org.Mm.egSYMBOL2EG))
setkey(eg, gene_id)
ensembleg <- as.data.table(as.data.frame(org.Mm.egENSEMBL))
setkey(ensembleg, gene_id)
fdat <- ensembleg[eg,,nomatch=0]
setkey(fdat, symbol)
```

Now join with symbols provided.  Some symbols map to multiple ensembl ids--we'll just take the first one listed.
We define the **Ranked** set of genes to be genes with cell cycle annotation, which BNC used to calculate the scLVM factor.  The **Unranked** set of genes are all others.
```{r}
mescFdat <- fdat[colnames(mesc),,mult='first']
mescFdat <- mescFdat[,ccRanked:=factor((ensembl_id %in% ccgenes), labels=c('FALSE'='Unranked', 'TRUE'='Ranked'))]
mescFdatDup <- fdat[colnames(mesc)]
```
This affects `r nrow(mescFdatDup)-nrow(mescFdat)` records.

## Loading mouse T-cell data
We repeat the process as above.
```{r loadt}
## Supplementary Data 1, sheet 1
T_cell_raw <- fread("data/T_cell_uncorrected.csv")
## sheet 2
T_cell_corrected <- fread("data/T_cell_corrected.csv") 
setnames(T_cell_raw, "V1", "cell_id")
setnames(T_cell_corrected, "V1", "cell_id")
## Supplementary Data 1, sheet 3
cluster <- fread("data/T_cell_cluster.csv")
setnames(cluster, 'Gata3HighCLuster', 'clusterid')
cluster <- cluster[,clusterid:=factor(clusterid)]
## Again, we generated using authors' code from https://github.com/PMBio/scLVM
lv_factor <- scan("data/X_ARD_1.txt")
```

Two gene symbols are duplicated, we'll just take the first one listed.
```{r}
T_cell_matrix <- as.matrix(T_cell_raw[,-1,with=FALSE])
T_cell_matrix_corrected <- as.matrix(T_cell_corrected[,-1,with=FALSE])
# Remove duplicated gene names (only 2)
T_cell_matrix <- T_cell_matrix[,unique(colnames(T_cell_matrix))]
T_cell_matrix_corrected <- T_cell_matrix_corrected[,unique(colnames(T_cell_matrix_corrected))]
stopifnot(all(colnames(T_cell_matrix_corrected)==colnames(T_cell_matrix)))

TcellCdat <- data.frame(cluster, panama=lv_factor, geomsizeUncorr=rowSums(T_cell_matrix))
```
Again, we'll need to map symbols to ensembl.
```{r}
TcellFdat <- fdat[colnames(T_cell_matrix),,mult='first']
TcellFdat <- TcellFdat[,ccRanked:=factor((ensembl_id %in% ccgenes), labels=c('FALSE'='Unranked', 'TRUE'='Ranked'))]
TcellFdatDup <- fdat[colnames(T_cell_matrix)]
```

There are `r nrow(mesc)` cells and `r ncol(mesc)` genes in the mESC data and
`r nrow(T_cell_matrix)` cells and `r ncol(T_cell_matrix)` genes in the T-cell data.


## $R^2$ attributable to Cell Cycle in mESC
First, we explore how much of the variance in the mESC data set can be explained by cell cycle.

Below, we define the $R^2$ as a ratio of linear model deviances (sums of squares) in a full and null model.
```{r}
lrTestLM <- function(Fit, drop){
    nFit <- update(Fit, as.formula(paste0('. ~ . -', drop)))
    lambda <- (deviance(nFit)-deviance(Fit))
    nDF <- (Fit$rank-nFit$rank)
    data.frame(lambda=lambda, P=1-pchisq(lambda, nDF), pctDev=pmin(lambda/deviance(nFit)*100))
    }
```
This function just compares the residual squared error with and without the factor `drop` (in our case, this will always be cell cycle), and calculates this change (`lambda`) and its ratio over the null deviance as a percentage.

BNC offer an adjustment in their variance decompositions for a factor that reflects technical variability.  We didn't attempt to replicate their calculation of this factor.  Below we compare the unadjusted values to some benchmark values from their paper and find reasonable agreement nonetheless, suggesting that these findings are robust to adjustment for putative sources of technical variability.

We consider regressions of `cycle_class` on the uncorrected log expression values.
```{r ccR2}
afit <- lm(mesc~cycle_class, data=mescCdat)
atest <- lrTestLM(afit, 'cycle_class')
P <- atest$P 
pctDev <- atest$pctDev
lambda <- atest$lambda
devModels <- data.table(model='cycle_class', symbol=colnames(coef(afit)), P, lambda, pctDev)
setkey(devModels, symbol)
setkey(mescFdat, symbol) 
devModels <- devModels[mescFdat,,allow.cartesian=TRUE]
```
We have joined the linear model results to the gene annotation data here.

Next, we report the variance estimates in median/90% gene in ranked/unranked sets, which gives the following result:
```{r tabDeviance, results='asis'}
devSum <- devModels[,{ 
    Q <- quantile(pctDev, prob=c(.5, .9), na.rm=TRUE)
    list('50%'=Q[1], '90%'=Q[2], Ngenes=.N)
}, keyby=list(ccRanked)]

knitr::kable(devSum)
```

*Table 1: Variance attributable to cell cycle in median/90% ranked and unranked genes*

### Comparing to Supplementary Figure 4b
We sought to compare our estimate to a description of these quantities in the paper.  We located an estimate of the effect of cell cycle by reading numbers off of figures S4b/S4d.

From this figure, fewer than 22% (1-7638/9571) of genes have >10% of variance attributable to cell cycle via "gold standard" including both ranked and unranked genes.
We reproduce the rest of the numbers from reading the gene counts annotated above the box plots.
Although this ignores the difference between ranked/unranked genes, we see that  BNC's result lies somewhere between the results for ranked/unranked genes.
```{r varianceEstimatesFigureS4}
## right endpoints of R^2 bins
R2 <- c(0, 10, 20, 30, 40, 50, 90)
## number of genes in bin
counts <- c(0, 7638, 1079, 483, 225, 97, 49)
## number of genes in bin and all bins below
countsBelow <- cumsum(counts)
figureData <- data.frame(R2, countsBelow, pctBelow=countsBelow/max(countsBelow), source='From Figure S4b')
```

```{r}
ggplot()+stat_ecdf(data=devModels, mapping=aes(x=pctDev, col=ccRanked)) +geom_line(data=figureData, mapping=aes(x=R2, y=pctBelow, col=source)) + xlab('% R^2 due to cell cycle') + ylab('Cumulative Distribution') + scale_color_discrete("Source") + theme(legend.position=c(.7, .3)) + geom_abline(aes(intercept=c(.5, .9), slope=0), lty=2)
```

*Figure 1: Cell cycle variability estimates*

#### Remark
Since there is no manifest variable that explains the "technical" variance that BNC subtract in their calculations, we think that it makes more sense to consider the undeflated estimates, since these represent the gene expression values that experimenters must use in practice, until a factor can be identified as a manifest variable for the technical variability.

## PCA
### PCA On Uncorrected mESC
PCA on all genes shows that the geometric size `geomsize` is principal component 1, explaining 9%-29% of the variance.
```{r}
getPCA <- function(exprMat, cData, scale.=FALSE){
    pcOut <- prcomp(exprMat, scale.=scale.)
    geomsize <- rowSums(exprMat)
    pcall <- data.frame(pcOut$x[,1:3], geomsize, cData)
    varExp <- pcOut$sdev^2/sum(pcOut$sdev^2)
    list(pcPairs=pcall, varianceExplained=varExp)
}

mescPca <- getPCA(mesc, mescCdat[,c('cycle_class', 'panama')])

```

We see that PC1 one is highly correlated with geometric size (`geomsize`) and the `panama` factor that we estimated.
We also observe that the cell cycle phase of a cell is correlated with `geomsize`/PC1 and discuss that point at greater length below.
```{r plotmescpca, fig.width=10, fig.height=10, dev=c('png', 'pdf')}
ggpairs(mescPca$pcPairs, colour='cycle_class')

```

*Figure 2: PCA and geometric library size in mESC experiment*

### PCA on Uncorrected Tcells
In the Tcells, we again see that PC1 one is highly correlated with geometric size (`geomsize`) and the `panama`  factor.
```{r plotTcpca, fig.width=10, fig.height=10}
tcpca <- getPCA(T_cell_matrix, TcellCdat[,c('panama', 'clusterid')])
ggpairs(tcpca$pcPairs, columns=c(1:5), colour='clusterid')

```

*Figure 3: PCA and geometric library size in Tcell experiment*


## Percentage variance due to PC1
```{r, results='asis'}
varExplained <- data.frame('data set'=c('mESC', 'T-cell'),
                           'PC1 Variance Explained'=c(mescPca$varianceExplained[1],
                               tcpca$varianceExplained[1]), check.names=F)
knitr::kable(varExplained) 
```

*Table 2: PCA PC1 variance explained*

### Remark on T-cell clustering
The T-cell clustering reported in the corrected T-cell data appears to depend substantially on the dimension reduction method used.  We are unable to replicate the clustering in the corrected or uncorrected data using linear PCA, or non-linear t-stochastic neighborhood embedding.  As the exact software version and parameters BNC used for the 'non-linear PCA' were not provided, we didn't attempt to apply that algorithm.

Using ordinary PCA, we found instead that the GATA3 `cluster` variable corresponds roughly to a half-space defined by PC2, but since multiple modes are not evident, we would not detect distinct clusters, but rather a gradient of cells.
```{r PCAforClusters}
tcCorrectedPCA <- getPCA(T_cell_matrix_corrected, TcellCdat[,c('panama', 'clusterid')])
qp <- ggplot(data = tcCorrectedPCA$pcPairs, aes(x = PC1, y = PC2)) + geom_density2d(color = "black",show_guide=FALSE) + geom_point(data = tcCorrectedPCA$pcPairs, aes(size = geomsize, color = clusterid))+scale_color_discrete("cluster id")+scale_size_continuous("geometric size")
qp1 <- qp + ggtitle('PCA Corrected')
qp2 <- qp %+% tcpca$pcPairs + ggtitle('PCA Uncorrected')
grid.arrange(qp1, qp2)

```

*Figure 4: Looking for clusters in Tcells using PCA on corrected and uncorrected data*

We also try to do the same using `t-SNE` as implemented in the `Rtsne` package, which is a non-linear dimension reduction technique, which has been successfully applied to single-cell data.

The figures above (PCA) and below (t-SNE) do now show any obvious clustering of the cells either before or after correction according to the clusters reported in BNC. An obvious grouping of the cells by geomsize is present.
```{r TSNEforClusters, results='hide'}
#exact tsne, set theta=0
tsne <- data.frame(tsne=Rtsne(T_cell_matrix, dims = 2, initial_dims = 50, perplexity = 20, theta = 0.0, check_duplicates = TRUE, pca = TRUE)$Y, geomsize=rowSums(T_cell_matrix))

tsneCorr <- data.frame(tsne=Rtsne(T_cell_matrix_corrected, dims = 2, perplexity = 20, theta = 0.0, check_duplicates = TRUE, pca = TRUE)$Y, geomsize=rowSums(T_cell_matrix_corrected)) 
qptsne <- qplot(data=cbind(tsne, TcellCdat), x=tsne.1, y=tsne.2, col=clusterid)+geom_density2d(aes(group=1),col="black",show_guide=FALSE)+geom_point(aes(size=geomsize))+scale_color_discrete("cluster id")+scale_size_continuous("geometric size")
qp3 <- qptsne %+% cbind(tsneCorr, TcellCdat) + ggtitle('TSNE Corrected')
qp4 <- qptsne + ggtitle('TSNE Uncorrected')
grid.arrange(qp3, qp4)
```

*Figure 5: Looking for clusters in Tcells using t-SNE on corrected and uncorrected data*



## Relationship between `geomsize`, `panama` and cell cycle
From the plots of the PCA in mESC, we observe that average geometric size and average panama factor both depend on the cell cycle.  Here we formally consider how much variance cell cycle explains in these factors by comparing the $R^2$ due to cell cycle on `panama` versus the $R^2$ due to geometric size, within each cell cycle.
```{r libsizepanama, results='asis'}
models <- c('panama ~ cycle_class', 'panama ~ geomsizeUncorr', 'panama ~ geomsizeUncorr', 'panama ~ geomsizeUncorr')
Subset <- list('all'=c('G1', 'S', 'G2M'), G1='G1', S='S', G2M='G2M')
R2 <- sapply(seq_along(models), function(i){
    l <- lm(as.formula(models[i]), data=mescCdat, subset=cycle_class %in% Subset[[i]])
    summary(l)$r.squared
})

knitr::kable(data.frame(Model=models, 'R^2' = R2, Subset=names(Subset), check.names=FALSE), row.names=FALSE)
```

*Table 3: dependence of panama factor on cell cycle and geometric size*

### Comparison of discrimant power of `geomsize` and `panama`
The comparison can also be inverted.  Using proportional odds logistic regression to classify cells by their geometric size and/or panama factor, we find modestly better performance using `panama`.  In both cases, the error rate exceeds 25%.  The S phase cells are especially resistant to classification.
```{r libsizepanamapolr}
## scale geomsize to prevent unfriendly numerics
pgeom <- polr(cycle_class ~ scale(geomsizeUncorr), data=mescCdat)
ppanama <- update(pgeom, . ~  panama)
tgeom <- table(predict(pgeom), mescCdat$cycle_class)
tpanama <- table(predict(ppanama), mescCdat$cycle_class)
```
`geomsize` has a error rate of `r 1-sum(diag(tgeom))/sum(tgeom)` and confusion matrix:
```{r, results='asis'}
knitr::kable(tgeom)
```

*Table 4: cell cycle classification using geometric size*


`panama` has a error rate of `r 1-sum(diag(tpanama))/sum(tpanama)` and confusion matrix:
```{r, results='asis'}
knitr::kable(tpanama)
```

*Table 5: cell cycle classification using panama*


The heavy dependence of `panama` on geometric size means that the classificiation power of panama relies on geometric size being a reasonable instrument for cell cycle, which is an assumption that has not been thoroughly tested.

## Gene Set Enrichment Analysis using Corrected Data
To conclude our analysis, we investigated if the scLVM factor would be successful in eliminating cell cycle signal when testing for differences between the T-cell clusters.

We considered gene set enrichment analysis on the corrected T-cell data.
```{r gsea, results='hide'}
eSet <- ExpressionSet(t(T_cell_matrix_corrected))
pData(eSet) <- TcellCdat
```
We used the Broad Institute's "Reactome" module available here: http://www.broadinstitute.org/gsea/msigdb/collections.jsp.
```{r}
c2_set <- getGmt("data/c2.cp.reactome.v4.0.symbols.gmt")
gene_ids <- geneIds(c2_set)
## # Camera requires gene-indices
design <- model.matrix(~clusterid, eSet)
sets_indices <- ids2indices(gene_ids, toupper(rownames(eSet))) 
res <- camera(eSet, sets_indices, design=design)
```

We'll define the following keywords as relating to cell cycle: cell cycle, cycling/cyclin (preceded by a word boundary), M, S, G0/1/2 (preceded and followed by word boundaries), replication/replicative, mitotic, mitosis, chrom (preceded by a word boundary).
```{r}
cycleKw <- '(CELL_CYCLE)|(_CYCLIN)|((_|^)M_)|((_|^)S_)|((_|^)G[012]+_)|REPLICATI|MITOTIC|MITOSIS|(_|^)CHROM'

```

```{r topgenes, results='asis'}
res$Direction <- ifelse(res$Direction=='Up', 1, -1)
res$set <- row.names(res)
setDT(res)
setkey(res, FDR) 
res <- res[,':='(totalrank=rank(PValue))]
res <- res[,cycleKw:=str_detect(set, cycleKw)]
resFDR <- res[FDR<.05,]
nFDR <- nrow(resFDR)
```
In the `r nrow(resFDR)` modules with a FDR-q value 5% or less, a substantial number relate to cell cycle.

Among the top 20 modules, there is little suggestion of enrichment for immue-related modules.  Many modules relate to cell cycling.
```{r, results='asis'}
knitr::kable(resFDR[1:20,list(Direction, FDR, totalrank, substr(str_replace(set, 'REACTOME_', ''), 1, 40), cycleKw)])  
```

*Table 6: gene set enrichment analysis using reactome modules on T cell clusters*

The same is true when using the GO Biological Process modules. All 20 top modules are directly or indirectly (DNA repair or DNA replication) cell-cycle associated.

```{r gobp}
c5_set <- getGmt("data/c5.bp.v4.0.symbols.gmt")
gene_ids <- geneIds(c5_set)
## # Camera requires gene-indices
design <- model.matrix(~clusterid, eSet)
sets_indices <- ids2indices(gene_ids, toupper(rownames(eSet))) 
res <- camera(eSet, sets_indices, design=design)
```

```{r topgenes_gobp, results='asis'}
res$Direction <- ifelse(res$Direction=='Up', 1, -1)
res$set <- row.names(res)
setDT(res)
setkey(res, FDR) 
res <- res[,':='(totalrank=rank(PValue))]
res <- res[,cycleKw:=str_detect(set, cycleKw)]  
resFDR <- res[FDR<.05,] 
nFDR <- nrow(resFDR)
```

```{r gobp_result, results='asis'}
knitr::kable(resFDR[1:20,list(Direction, FDR, totalrank, set, cycleKw)])  
```

*Table 7: gene set enrichment analysis using GO-biological process modules on T cell clusters*

### Module Diversity
We considered if cell cycle annotation was over-represented in the Reactome or GO Biological Process modules.  The following function counts modules with any mention of keywords noted above, and then finds the overlap between genes defined in the keyword-containing modules and all other modules.
```{r module_diversity}
getCCMods <- function(set){
    mods <-  data.table(melt(geneIds(set), value.name='geneid'))
    setnames(mods, 'L1', 'module')
    mods[,':='(isCC=str_detect(module, cycleKw),
                   geneid=as.character(geneid))]
    mods[,size:=.N,keyby=module]
    nmods = mods[,.(nmods = length(unique(module))),keyby=isCC]
    genes <- mods[,.(ccGene=any(isCC), .N),keyby=list(geneid)]
    mods <- merge(genes, mods,by='geneid')
    modsccgenes <- dcast.data.table(mods, module+isCC +size~ ccGene)
    modsccgenes[,ccpct:=`TRUE`/size]
    medianOverlap <- modsccgenes[,.(median=median(ccpct)),keyby=list(isCC)]
    list(ccMod=nmods[isCC==TRUE,nmods],
         totalMod=nmods[,sum(nmods)],
         overlapCCPct=round(medianOverlap[isCC==FALSE,median]*100, 0), modsccgenes=modsccgenes)
    }

reactome <- getCCMods(c2_set)
bp <- getCCMods(c5_set)
```

In the Reactome modules only `r reactome$ccMod` (out of `r reactome$totalMod`) modules had a keyword reflecting cell cycle. In the non-cycle set, the median non-annotated module included only `r reactome$overlapCCPct`% of genes with any annotation.

In the GO-Biological Process modules these statistics were `r bp$ccMod`/`r bp$totalMod` and `r bp$overlapCCPct`%, respectively.


## Session Info
This report was generated using
```{r}
sessionInfo()
```
as a pdf from source code available at https://github.com/RGLab/BNCResponse/blob/master/BNCAnalysis.Rmd .
